packages/knit-gateway/src/json.ts

```diff
@@ -0,0 +1,166 @@
+// Copyright 2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import type {
+  AnyMessage,
+  FieldInfo,
+  IMessageTypeRegistry,
+  JsonObject,
+  JsonValue,
+  MessageType,
+  PlainMessage,
+} from "@bufbuild/protobuf";
+import type {
+  Schema,
+  Schema_Field,
+  Schema_Field_Type,
+  Schema_Field_Type_MapType,
+  Schema_Field_Type_RepeatedType,
+} from "@buf/bufbuild_knit.bufbuild_es/buf/knit/gateway/v1alpha1/knit_pb.js";
+import { wktSet } from "./wkt.js";
+import type { Relation } from "./gateway.js";
+import { Code, ConnectError } from "@bufbuild/connect";
+
+export interface Patch {
+  base: AnyMessage;
+  target: JsonObject;
+  field: PlainMessage<Schema_Field> & {
+    relation: Relation;
+    params?: AnyMessage;
+  };
+}
+
+/**
+ * Formats the message as a `JsonValue` according to the schema. It also returns the
+ * patches to be applied to the message.
+ *
+ * It doesn't verify if the json object is valid for the schema, it just tries to apply
+ * the mask. It will throw an error for some mismatches like Array instead of object and doesn't
+ * throw for some like string instead of number.
+ *
+ * @internal
+ */
+export function formatMessage(
+  message: AnyMessage,
+  schema: PlainMessage<Schema>,
+  typeRegistry?: IMessageTypeRegistry
+): [JsonValue, Patch[]] {
+  const type = message.getType();
+  if (wktSet.has(type.typeName))
+    return [message.toJson({ typeRegistry: typeRegistry }), []];
+  const formattedObject: JsonObject = {};
+  const patches: Patch[] = [];
+  for (const field of schema.fields) {
+    const fieldInfo = type.fields.findJsonName(
+      field.jsonName === "" ? field.name : field.jsonName
+    );
+    if (fieldInfo === undefined) {
```

Comment:
It seems like we should omit this check and prioritize `if (field.relation !== undefined)`. Basically, if the schema says it's a relation, generate a patch. And and only if it's not a relation do we try to look up the field info by JSON name.